### PR TITLE
Add option to replay an account and selector.

### PR DESCRIPTION
**Pull request by Anandkumar26**
Add structs to capture acccount configuration and selector
configuration.
    
Account config helps to know whether an account is succesfully
initialized or not, whethere credentials are valid or not. Also
provides an option to retry errors when an cloud plug-in returns error.
Selector config keeps the latest version of selector.
    
This data, helps to replay all the selectors for a given account, when
account transitions for an non-initialized state to initialized
state, so that filters in cloud plug-in are re-configured with
latest version of selector.
    
Also move provider type to account config

### CODE REVIEW

File: pkg/accountmanager/manager.go [hunk 1]
```diff
@@ -257,3 +294,152 @@ func (a *AccountManager) getAccountPoller(name *types.NamespacedName) (*accountP
 	}
 	return accPoller, true
 }
+
+// getAccountConfig returns the account configuration.
+func (a *AccountManager) getAccountConfig(name *types.NamespacedName) *accountConfig {
+	a.mutex.RLock()
+	defer a.mutex.RUnlock()
+	config, ok := a.accountConfigMap[*name]
```
Comment (reachjainrahul) [inline pkg/accountmanager/manager.go:302]: use golang style lookup

File: pkg/accountmanager/manager.go [hunk 2]
```diff
@@ -257,3 +294,152 @@ func (a *AccountManager) getAccountPoller(name *types.NamespacedName) (*accountP
...
+
+// addSelectorConfig add the current selector configuration to the specified account.
+func (a *AccountManager) addSelectorConfig(accountNamespacedName, selectorNamespacedName *types.NamespacedName,
+	selector *crdv1alpha1.CloudEntitySelector) error {
+	acctConfig := a.getAccountConfig(accountNamespacedName)
+	if acctConfig == nil {
+		return fmt.Errorf("failed to get account config")
+	}
+	config, ok := acctConfig.selectorConfigMap[*selectorNamespacedName]
+	if !ok {
```
Comment (reachjainrahul) [inline pkg/accountmanager/manager.go:358]: same here.. golang style

File: pkg/accountmanager/manager.go [hunk 3]
```diff
@@ -257,3 +294,152 @@ func (a *AccountManager) getAccountPoller(name *types.NamespacedName) (*accountP
...
+	}
+	return config.retry
+}
+
+// replaySelectorsForAccount replays all the selectors that belong to this specified account,
+// set/resets the selector status.
+func (a *AccountManager) replaySelectorsForAccount(namespacedName *types.NamespacedName, config *accountConfig) {
+	for _, filterConfig := range config.selectorConfigMap {
+		a.Log.Info("Re-playing selector", "account", namespacedName, "selector", filterConfig.namespacedName)
+		_, err := a.AddResourceFiltersToAccount(namespacedName, filterConfig.namespacedName, filterConfig.selector, true)
```
Comment (reachjainrahul) [inline pkg/accountmanager/manager.go:422]: update filter directly by calling plugin.

File: pkg/accountmanager/poller.go [hunk 4]
```diff
@@ -126,8 +126,8 @@ func (p *accountPoller) addOrUpdateSelector(selector *crdv1alpha1.CloudEntitySel
 		}
 	}
 
-	// Populate selector specific fields in the accPoller created by CPA, needed for setting owner reference in VM CR.
-	p.selector = selector.DeepCopy()
+	// Store selector to filter cloud resources based on selector.
+	p.selector = selector
```
Comment (reachjainrahul) [inline pkg/accountmanager/poller.go:130]: continue using Deepcopy. This way, we release buffer of selector from reconciler

File: pkg/controllers/cloudentityselector/controller.go [hunk 5]
```diff
@@ -17,6 +17,8 @@ package cloudentityselector
 import (
 	"context"
 	"fmt"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
```
Comment (reachjainrahul) [inline pkg/controllers/cloudentityselector/controller.go:20]: sort imports

File: pkg/controllers/cloudentityselector/controller.go [hunk 6]
```diff
@@ -145,3 +151,23 @@ func (r *CloudEntitySelectorReconciler) processDelete(selectorNamespacedName *ty
 	_ = r.AccManager.RemoveResourceFiltersFromAccount(&accountNamespacedName, selectorNamespacedName)
 	return nil
 }
+
+// setStatus sets the status on the CloudEntitySelector CR.
+func (r *CloudEntitySelectorReconciler) setStatus(namespacedName *types.NamespacedName, err error) {
+	var errorMsg string
+	if err != nil {
```
Comment (reachjainrahul) [inline pkg/controllers/cloudentityselector/controller.go:158]: wont this be called if there is an error. If its nil, do you want to respond? Or you are using this to clear error?

File: pkg/accountmanager/manager.go [hunk 7]
```diff
@@ -43,72 +45,66 @@ const (
...
+	namespacedName *types.NamespacedName
+	providerType   common.ProviderType
+	// Indicates account initialization state.
+	accountInitState bool
+	// Indicates credentials are valid or not.
+	credentialsValid bool
+	// retry upon failure.
+	retry bool
+	// capture the filter configuration of each CloudEntitySelector.
+	// Lock is not added, since no two CloudEntitySelector CRs will access the same entry.
```
[the remark this answers is not in the document]
Comment (Anandkumar26) [reply pkg/accountmanager/manager.go]: Done

File: pkg/accountmanager/manager.go [hunk 8]
```diff
@@ -43,72 +45,66 @@ const (
...
-func (a *AccountManager) addAccountProviderType(namespacedName *types.NamespacedName,
-	provider runtimev1alpha1.CloudProvider) common.ProviderType {
-	a.mutex.Lock()
-	defer a.mutex.Unlock()
-
-	providerType := common.ProviderType(provider)
-	a.accountProviderType[*namespacedName] = providerType
-	return providerType
+type selectorConfig struct {
+	namespacedName *types.NamespacedName
```
[the remark this answers is not in the document]
Comment (Anandkumar26) [reply pkg/accountmanager/manager.go]: When I stared, I had more information in this structure. 
After cleaning up, its not required.

File: pkg/accountmanager/manager.go [hunk 9]
```diff
@@ -121,57 +117,82 @@ func (a *AccountManager) AddAccount(namespacedName *types.NamespacedName, accoun
 	} else {
 		accPoller.restartPoller(namespacedName)
 	}
-	return nil
+
+	if !config.accountInitState {
```
[the remark this answers is not in the document]
Comment (Anandkumar26) [reply pkg/accountmanager/manager.go]: Handled it.

File: pkg/accountmanager/manager.go [hunk 10]
```diff
@@ -121,57 +117,82 @@ func (a *AccountManager) AddAccount(namespacedName *types.NamespacedName, accoun
...
+		// Replay CES CR only when account init state is changed from failure to success.
+		a.replaySelectorsForAccount(namespacedName, config)
+		// Set account init state to true as there were no errors reported from cloud plug-in.
+		config.retry = false
```
[the remark this answers is not in the document]
Comment (Anandkumar26) [reply pkg/accountmanager/manager.go]: Done

File: pkg/accountmanager/manager.go (hunk 1 above)
Comment (Anandkumar26) [reply pkg/accountmanager/manager.go:302]: config will be used later in the function.

File: pkg/accountmanager/manager.go (hunk 3 above)
Comment (Anandkumar26) [reply pkg/accountmanager/manager.go]: We would also need to update the selectors in the poller. Doing so, we will end having another version of the AddResourceFiltersToAccount() function. So avoiding it.

File: pkg/accountmanager/poller.go (hunk 4 above)
Comment (Anandkumar26) [reply pkg/accountmanager/poller.go:130]: Updated the account config structure to use deepcopy. Poller can operate on the selector stored in the account config.

File: pkg/controllers/cloudentityselector/controller.go (hunk 5 above)
Comment (Anandkumar26) [reply pkg/controllers/cloudentityselector/controller.go]: Done

File: pkg/accountmanager/manager.go [hunk 11]
```diff
@@ -43,139 +47,155 @@ const (
...
-	accPollers          map[types.NamespacedName]*accountPoller
-	accountProviderType map[types.NamespacedName]common.ProviderType
+	mutex            sync.RWMutex
+	Inventory        inventory.Interface
+	accPollers       map[types.NamespacedName]*accountPoller
+	accountConfigMap map[types.NamespacedName]*accountConfig
+}
+
+type accountConfig struct {
+	namespacedName *types.NamespacedName
```
Comment (reachjainrahul) [inline pkg/accountmanager/manager.go:68]: nit: This is not needed. You can clean it up in later.

**Review by reachjainrahul — APPROVED** [review]
/LGTM

**Pull request merged**